core/src/banking_stage.rs

```diff
@@ -985,23 +1049,48 @@ impl BankingStage {
         msgs: &Packets,
         transaction_indexes: &[usize],
         secp256k1_program_enabled: bool,
-    ) -> (Vec<HashedTransaction<'static>>, Vec<usize>) {
-        transaction_indexes
+        cost_model: &Arc<CostModel>,
+        cost_tracker: &Arc<Mutex<CostTracker>>,
+    ) -> (Vec<HashedTransaction<'static>>, Vec<usize>, Vec<usize>) {
+        // make a copy of cost_tracker to statically check limit for this batch of packets
+        let guard = cost_tracker.lock().unwrap();
+        let mut cost_tracker = guard.clone();
+        drop(guard);
+
+        let mut retryable_transaction_packet_indexes: Vec<usize> = vec![];
+        let (filtered_transactions, filter_transaction_packet_indexes) = transaction_indexes
             .iter()
             .filter_map(|tx_index| {
                 let p = &msgs.packets[*tx_index];
                 let tx: Transaction = limited_deserialize(&p.data[0..p.meta.size]).ok()?;
                 if secp256k1_program_enabled {
                     tx.verify_precompiles().ok()?;
                 }
+
+                // get transaction cost via immutable cost_model; try to add cost to tracker
+                // synchronizlysynchronously. Check result before continue;
+                let tx_cost = cost_model.calculate_cost(&tx);
+                let result = cost_tracker.try_add(tx_cost);
+                if result.is_err() {
+                    debug!("transaction {:?} would exceed limit: {:?}", tx, result);
+                    retryable_transaction_packet_indexes.push(*tx_index);
+                    return None;
+                }
+
                 let message_bytes = Self::packet_message(p)?;
                 let message_hash = Message::hash_raw_message(message_bytes);
                 Some((
                     HashedTransaction::new(Cow::Owned(tx), message_hash),
                     tx_index,
                 ))
             })
-            .unzip()
+            .unzip();
+
```

Comment:
I think we might want to just do another loop after doing the hash and secp256k1, so we do those outside the Mutex lock.

Reply:
added comment to document the use of local copy of tracker, no lock here